File: mcp-sentinel-python/src/mcp_sentinel/models/scan_result.py
"""
Scan result model for aggregating vulnerability findings.
"""

from typing import List, Dict, Any, Optional
from datetime import datetime
from pydantic import BaseModel, Field

from mcp_sentinel.models.vulnerability import Vulnerability, Severity


class ScanStatistics(BaseModel):
    """Statistics about the scan."""

    total_files: int = 0
    scanned_files: int = 0
    total_vulnerabilities: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    info_count: int = 0
    scan_duration_seconds: float = 0.0
    engines_used: List[str] = Field(default_factory=list)


class ScanResult(BaseModel):
    """
    Result of a security scan.

    Attributes:
        scan_id: Unique scan identifier
        target: Path or URL that was scanned
        vulnerabilities: List of found vulnerabilities
        statistics: Scan statistics
        config: Configuration used for scan
        started_at: Scan start time
        completed_at: Scan completion time
        status: Scan status
        error: Error message if scan failed
    """

    scan_id: str = Field(default_factory=lambda: f"scan-{datetime.utcnow().timestamp()}")
    target: str
    vulnerabilities: List[Vulnerability] = Field(default_factory=list)
    statistics: ScanStatistics = Field(default_factory=ScanStatistics)
    config: Dict[str, Any] = Field(default_factory=dict)

    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None
    status: str = "pending"  # pending, running, completed, failed
    error: Optional[str] = None

    def add_vulnerability(self, vuln: Vulnerability) -> None:
        """Add a vulnerability to the result."""
        self.vulnerabilities.append(vuln)
        self.statistics.total_vulnerabilities += 1

        # Update severity counts
        if vuln.severity == Severity.CRITICAL:
            self.statistics.critical_count += 1
        elif vuln.severity == Severity.HIGH:
            self.statistics.high_count += 1
        elif vuln.severity == Severity.MEDIUM:
            self.statistics.medium_count += 1
        elif vuln.severity == Severity.LOW:
            self.statistics.low_count += 1
        else:
            self.statistics.info_count += 1

    def get_by_severity(self, severity: Severity) -> List[Vulnerability]:
        """Get all vulnerabilities of a specific severity."""
        return [v for v in self.vulnerabilities if v.severity == severity]

    def get_by_file(self, file_path: str) -> List[Vulnerability]:
        """Get all vulnerabilities in a specific file."""
        return [v for v in self.vulnerabilities if v.file_path == file_path]

    def has_critical_findings(self) -> bool:
        """Check if scan has critical vulnerabilities."""
        return self.statistics.critical_count > 0

    def has_high_or_critical(self) -> bool:
        """Check if scan has high or critical vulnerabilities."""
        return (self.statistics.critical_count + self.statistics.high_count) > 0

    def risk_score(self) -> float:
        """
        Calculate overall risk score (0-100).

        Weighted average based on severity distribution.
        """
        if self.statistics.total_vulnerabilities == 0:
            return 0.0

        total_score = sum(v.risk_score() for v in self.vulnerabilities)
        return total_score / self.statistics.total_vulnerabilities

    def summary(self) -> str:
        """Generate a text summary of the scan results."""
        return (
            f"Scan {self.scan_id}\n"
            f"Target: {self.target}\n"
            f"Status: {self.status}\n"
            f"Duration: {self.statistics.scan_duration_seconds:.2f}s\n"
            f"Files Scanned: {self.statistics.scanned_files}/{self.statistics.total_files}\n"
            f"Total Vulnerabilities: {self.statistics.total_vulnerabilities}\n"
            f"  Critical: {self.statistics.critical_count}\n"
            f"  High: {self.statistics.high_count}\n"
            f"  Medium: {self.statistics.medium_count}\n"
            f"  Low: {self.statistics.low_count}\n"
            f"  Info: {self.statistics.info_count}\n"
            f"Risk Score: {self.risk_score():.1f}/100"
        )

    class Config:
        json_schema_extra = {
            "example": {
                "scan_id": "scan-1234567890",
                "target": "/path/to/project",
                "vulnerabilities": [],
                "statistics": {
                    "total_files": 100,
                    "scanned_files": 95,
                    "total_vulnerabilities": 15,
                    "critical_count": 2,
                    "high_count": 5,
                    "medium_count": 6,
                    "low_count": 2,
                    "info_count": 0,
                    "scan_duration_seconds": 7.8,
                    "engines_used": ["static", "semantic", "ai"]
                },
                "status": "completed"
            }
        }

File: rust-legacy/src/models/scan_result.rs
//! Scan result data model

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::vulnerability::{Severity, Vulnerability};

/// Summary statistics for scan results
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ScanSummary {
    pub total_issues: usize,
    pub critical: usize,
    pub high: usize,
    pub medium: usize,
    pub low: usize,
    pub risk_score: u8, // 0-100
}

impl ScanSummary {
    /// Create a summary from a list of vulnerabilities
    pub fn from_vulnerabilities(vulnerabilities: &[Vulnerability]) -> Self {
        let critical = vulnerabilities
            .iter()
            .filter(|v| v.severity == Severity::Critical)
            .count();
        let high = vulnerabilities
            .iter()
            .filter(|v| v.severity == Severity::High)
            .count();
        let medium = vulnerabilities
            .iter()
            .filter(|v| v.severity == Severity::Medium)
            .count();
        let low = vulnerabilities
            .iter()
            .filter(|v| v.severity == Severity::Low)
            .count();

        // Risk score calculation: weighted by severity
        // Critical: 40 points, High: 20 points, Medium: 5 points, Low: 1 point
        // Capped at 100
        let risk_score = ((critical * 40 + high * 20 + medium * 5 + low).min(100)) as u8;

        Self {
            total_issues: vulnerabilities.len(),
            critical,
            high,
            medium,
            low,
            risk_score,
        }
    }
}

/// Metadata about the scan
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ScanMetadata {
    pub scan_duration_ms: u64,
    pub engines_used: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub llm_provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub llm_model: Option<String>,
}

/// Complete scan result
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ScanResult {
    /// Version of MCP Sentinel that performed the scan
    pub version: String,

    /// Unique scan identifier
    pub scan_id: String,

    /// Timestamp when scan started
    pub timestamp: DateTime<Utc>,

    /// Target that was scanned (path, URL, etc.)
    pub target: String,

    /// Engines used for scanning
    pub engines: Vec<String>,

    /// Summary statistics
    pub summary: ScanSummary,

    /// List of detected vulnerabilities
    pub vulnerabilities: Vec<Vulnerability>,

    /// Scan metadata
    pub metadata: ScanMetadata,
}

impl ScanResult {
    /// Create a new scan result
    pub fn new(target: impl Into<String>, engines: Vec<String>) -> Self {
        Self {
            version: crate::VERSION.to_string(),
            scan_id: Uuid::new_v4().to_string(),
            timestamp: Utc::now(),
            target: target.into(),
            engines,
            summary: ScanSummary {
                total_issues: 0,
                critical: 0,
                high: 0,
                medium: 0,
                low: 0,
                risk_score: 0,
            },
            vulnerabilities: Vec::new(),
            metadata: ScanMetadata {
                scan_duration_ms: 0,
                engines_used: Vec::new(),
                llm_provider: None,
                llm_model: None,
            },
        }
    }

    /// Add a vulnerability to the result
    pub fn add_vulnerability(&mut self, vuln: Vulnerability) {
        self.vulnerabilities.push(vuln);
        self.update_summary();
    }

    /// Add multiple vulnerabilities
    pub fn add_vulnerabilities(&mut self, vulns: Vec<Vulnerability>) {
        self.vulnerabilities.extend(vulns);
        self.update_summary();
    }

    /// Update summary statistics based on current vulnerabilities
    fn update_summary(&mut self) {
        self.summary = ScanSummary::from_vulnerabilities(&self.vulnerabilities);
    }

    /// Filter vulnerabilities by minimum severity
    pub fn filter_by_severity(&self, min_severity: Severity) -> Vec<&Vulnerability> {
        self.vulnerabilities
            .iter()
            .filter(|v| v.severity >= min_severity)
            .collect()
    }

    /// Check if scan found any issues at or above a severity level
    pub fn has_issues_at_level(&self, min_severity: Severity) -> bool {
        self.vulnerabilities
            .iter()
            .any(|v| v.severity >= min_severity)
    }

    /// Set scan duration
    pub fn set_duration(&mut self, duration_ms: u64) {
        self.metadata.scan_duration_ms = duration_ms;
    }

    /// Set LLM provider info
    pub fn set_llm_info(&mut self, provider: impl Into<String>, model: impl Into<String>) {
        self.metadata.llm_provider = Some(provider.into());
        self.metadata.llm_model = Some(model.into());
    }

    /// Get severity badge text
    pub fn severity_badge(&self) -> &'static str {
        match self.summary.risk_score {
            90..=100 => "🔴 CRITICAL",
            70..=89 => "🟠 HIGH",
            40..=69 => "🟡 MEDIUM",
            _ => "🔵 LOW",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::vulnerability::{Vulnerability, VulnerabilityType};

    #[test]
    fn test_scan_summary_calculation() {
        let vulns = vec![
            Vulnerability::new(
                "C-001",
                VulnerabilityType::CommandInjection,
                Severity::Critical,
                "Test",
                "Desc",
            ),
            Vulnerability::new(
                "H-001",
                VulnerabilityType::CommandInjection,
                Severity::High,
                "Test",
                "Desc",
            ),
            Vulnerability::new(
                "M-001",
                VulnerabilityType::CommandInjection,
                Severity::Medium,
                "Test",
                "Desc",
            ),
        ];

        let summary = ScanSummary::from_vulnerabilities(&vulns);
        assert_eq!(summary.total_issues, 3);
        assert_eq!(summary.critical, 1);
        assert_eq!(summary.high, 1);
        assert_eq!(summary.medium, 1);
        assert_eq!(summary.low, 0);
        // 1*40 + 1*20 + 1*5 = 65
        assert_eq!(summary.risk_score, 65);
    }

    #[test]
    fn test_scan_result_add_vulnerabilities() {
        let mut result = ScanResult::new("test-target", vec!["static".to_string()]);

        result.add_vulnerability(Vulnerability::new(
            "C-001",
            VulnerabilityType::CommandInjection,
            Severity::Critical,
            "Test",
            "Desc",
        ));

        assert_eq!(result.vulnerabilities.len(), 1);
        assert_eq!(result.summary.total_issues, 1);
        assert_eq!(result.summary.critical, 1);
    }

    #[test]
    fn test_filter_by_severity() {
        let mut result = ScanResult::new("test-target", vec!["static".to_string()]);

        result.add_vulnerability(Vulnerability::new(
            "C-001",
            VulnerabilityType::CommandInjection,
            Severity::Critical,
            "Test",
            "Desc",
        ));
        result.add_vulnerability(Vulnerability::new(
            "L-001",
            VulnerabilityType::CommandInjection,
            Severity::Low,
            "Test",
            "Desc",
        ));

        let high_and_above = result.filter_by_severity(Severity::High);
        assert_eq!(high_and_above.len(), 1);

        let all = result.filter_by_severity(Severity::Low);
        assert_eq!(all.len(), 2);
    }
}
